### PR TITLE
Redundant modifier rule

### PR DESCRIPTION
**Pull request by tagantroy**
Basic implementation of #251 
@vanniktech and @Mauin can you help me with description of Issue?
#hacktoberfest

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+			it.accept(classVisitor)
+			it.acceptChildren(functionVisitor)
+		}
+	}
+
+	private inner class ClassVisitor : DetektVisitor() {
+		override fun visitClass(klass: KtClass) {
+			super.visitClass(klass)
+			if (klass.isExplicitlyPublic()) {
+				report(CodeSmell(issue, Entity.from(klass)))
```
Comment (Mauin) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt]: `report(CodeSmell(issue.copy(description = "${klass.name} is explicitly marked as public. Public is the default visibility for classes. The public modifier is redundant."), Entity.from(klass)))`

Comment (Mauin) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt]: Just to give this specific issue some more context.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtModifierListOwner
+import org.jetbrains.kotlin.psi.KtNamedFunction
+
+class RedundantModifierRule(config: Config = Config.empty) : Rule(config) {
```
Comment (Mauin) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt]: The Kotlin warnings call this one: "Redundant visibility modifier" (as there are other modifiers that don't affect visibility)

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+				report(CodeSmell(issue, Entity.from(klass)))
+			}
+		}
+	}
+
+	private inner class FunctionVisitor : DetektVisitor() {
+		override fun visitNamedFunction(function: KtNamedFunction) {
+			super.visitNamedFunction(function)
+			if (function.isExplicitlyPublicNotOverridden()) {
+				report(CodeSmell(issue, Entity.from(function)))
```
Comment (Mauin) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt]: As above, I'd give this a description of: `"${function.name} is explicitly marked as public. Functions are public by default so this modifier is redundant."`

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtModifierListOwner
+import org.jetbrains.kotlin.psi.KtNamedFunction
+
+class RedundantModifierRule(config: Config = Config.empty) : Rule(config) {
+	override val issue: Issue = Issue("RedundantModifierRule", Severity.Style, "TODO")
```
Comment (Mauin) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt]: We can remove the description alltogether if all places that report it set a more concrete description.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt (hunk 2 above)
Comment (tagantroy) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantModifierRule.kt]: Thanks for review! Done.

**Comment by tagantroy** [comment]
@arturbosch can you mark it as hacktoberfest?

**Comment by Mauin** [comment]
@tagantroy I labeled it as hacktoberfest for you. But I don't think that's necessary. Any PR that is not labelled as "invalid" will count towards Hacktoberfest.

**Comment by tagantroy** [comment]
@Mauin Thanks!

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantVisibilityModifierRule.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtClass
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.psi.KtModifierListOwner
+import org.jetbrains.kotlin.psi.KtNamedFunction
+import org.jetbrains.kotlin.psi.KtProperty
+
+class RedundantVisibilityModifierRule(config: Config = Config.empty) : Rule(config) {
+	override val issue: Issue = Issue("RedundantVisibilityModifierRule", Severity.Style)
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantVisibilityModifierRule.kt]: Please add a generic issue description. This one is shown in `sonar-kotlin` rule profiles page.

Comment (schalkms) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantVisibilityModifierRule.kt]: I would also merge and squash these changes then. The commit messages are too generic IMHO.

Comment (tagantroy) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantVisibilityModifierRule.kt]: I'll add description on weekend

Comment (tagantroy) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantVisibilityModifierRule.kt]: @arturbosch done!

Comment (arturbosch) [reply detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/RedundantVisibilityModifierRule.kt]: Thanks!

**Pull request merged**